02-peer-protocol.md

```diff
@@ -49,11 +68,468 @@ transaction is confirmed are also not persistent - until you know the script
+  - if it has the lowest total satoshis contributed, as defined
+    by total `tx_add_input` values:
+    - MUST transmit their `tx_signatures` before the peer
+  - MUST order the `witness_stack`s by the `serial_id` of the input they
+    correspond to
+  - number of `witness_stack`s MUST equal the number of inputs they added
+
+The receiving node:
+  - MUST fail the negotiation if:
+    - if they have not sent a `tx_signatures` message and they are the
+      peer with the lowest total satoshis contributed, as defined by
+      `tx_add_input` values
+    - the message contains an empty `witness_stack`
+    - the number of `witness_stack`s does not equal the number of inputs
+      added by the sending node
+    - the `txid` does not match the txid of the transaction
+  - SHOULD apply the `witness`es to the transaction and broadcast it
+  - MUST reply with their `tx_signatures` if not already transmitted
+
+#### Rationale
+
+The peer with the lowest total of inputs amounts must transmit its
+`tx_signatures` first. If both peers have contributed equal input amounts,
+the peer with the first `node_id` (sorted lexicographically) sends their
+`tx_signatures` first.  This gives in a strict ordering of transmission
+for any multiparty tx collaboration.
+
+`witness` is the data for a witness element in a witness stack.
+Witness elements should *not* include their length.
+
+Witness data must be sorted according to the `serial_id` of
+the corresponding input.
+
+While the `minimum fee` is calculated and verified at `tx_complete` conclusion,
+it is possible for the fee for the exchanged witness data to be underpaid.
+It is the responsibility of the sending peer to correctly account for the
+required fee, e.g. a multisig witness stack whose weight exceeds 107.
+
+### The `tx_init_rbf` Message
+
+This message initiates a replacement of the transaction after it's been
+completed.
+
+1. type: 72 (`tx_init_rbf`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u32`:`locktime`]
+   * [`u32`:`feerate`]
+   * [`tx_init_rbf_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `tx_init_rbf_tlvs`
+2. types:
+    1. type: 0 (`funding_output_contribution`)
+    2. data:
+        * [`tu64`:`satoshis`]
+
+#### Requirements
+
+The sender:
+  - MUST set `feerate` greater than or equal to 25/24 times the `feerate`
+    of the previously constructed transaction, rounded down.
+
+The recipient:
+  - MUST respond either by failing the negotiation or with `tx_ack_rbf`.
+  - MUST fail the negotiation if:
+    - the `feerate` is not greater than 25/24 times `feerate` of the last
+      successfully constructed transaction
+  - MAY fail the negotiation for any reason
+
+#### Rationale
+
+`feerate` is the feerate this transaction will pay. It must be at least
+1/24 greater than the last used `feerate`, rounded down to the nearest
+satoshi to ensure there is progress.
+
+E.g. if the last `feerate` was 520, the next sent `feerate` must be 541
+(520 * 25 / 24 = 541.667, rounded down to 541).
```

Comment:
this allows for creation of txs that cant be broadcast when the initial fee rate is below `24*minrelayfee` sat/vB

i think it also shouldn't be allowed to initiate rbf with a feerate that does not cover the retransmission cost
the retransmission cost is `minrelayfee` * `virtual size` rounded up in sats

minrelayfee is 1 sat/vB at the moment but can change in the future (`bitcoin-cli getmempoolinfo`, returns in BTC/kvB)

source:
https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#implementation-details number 4

Reply:
I don't think the lightning spec should take those details into account. If bitcoin changes relay policies, we don't want to have to change the spec, only implementations should change. IMO it's ok to be somewhat hand-wavy regarding RBF, stating something general like "implementations are free to reject any RBF attempt that may fail to propagate".

I think it's good that the spec requires a minimal % increase, as it's protects against DoS, but this is just a lower bound.

Also, it's not such a big deal if we generate a tx that doesn't propagate because it doesn't meet the bitcoin network's relay policies, we'll just RBF again after some time and eventually we'll have something that relays and confirms.

Reply:
> I don't think the lightning spec should take those details into account. If bitcoin changes relay policies, we don't want to have to change the spec

yeah i get that, `minrelayfee` could be put in the spec (as text, not 1 sat/vB), and even implementation dont have to change if the backend supplies this information. but this can get overly complicated because some implementation may not use a backend that supplies this info, and 2 nodes could have a different value for this

> IMO it's ok to be somewhat hand-wavy regarding RBF, stating something general like "implementations are free to reject any RBF attempt that may fail to propagate".
> 
> I think it's good that the spec requires a minimal % increase, as it's protects against DoS, but this is just a lower bound.
> 
> Also, it's not such a big deal if we generate a tx that doesn't propagate because it doesn't meet the bitcoin network's relay policies, we'll just RBF again after some time and eventually we'll have something that relays and confirms.

true

Reply:
> I think it's good that the spec requires a minimal % increase, as it's protects against DoS, but this is just a lower bound.
> 
> 
> 
> Also, it's not such a big deal if we generate a tx that doesn't propagate because it doesn't meet the bitcoin network's relay policies, we'll just RBF again after some time and eventually we'll have something that relays and confirms.

I agree with this. It's worth keeping in mind in any case that splicing (which is using the same protocol) may never have anything that confirms and has to maintain HTLC sig duplicates for every RBF. So the cost of RBF spam there is higher.

Reply:
The goal of splicing is lowest possible fees though so we want low fee attempts. Given that & it's negotiated between two peers who both bear the burden -- I suspect it's totally worth the extra overhead of low RBF bump minimums.